src/main/derived/ArbitraryDeriving.scala

```diff
@@ -9,48 +9,52 @@ import scala.compiletime.summonFrom
 import scala.compiletime.summonInline
 import scala.deriving.*
 
-/**
- * Derivation of `Arbitrary`s.
- */
-private trait ArbitraryDeriving:
+private case class Gens[+T](gens: List[Gen[T]]):
 
-  /** Helper for deriving `Gen`s for all subclasses of a "sum" (including sub-traits). */
-  private trait GensList[+T]:
-    def gens: List[Gen[T]]
+  private def list: List[Gen[T]] = gens
 
-  private trait LowPriorityGensListGivens {
+  def combine[U >: T](that: Gens[U]): Gens[U] =
+    Gens[U](this.gens ++ that.gens)
 
-    /** Fallback for anything that is not a sum. */
-    inline given nonSumGensList[T](using a: Arbitrary[T]): GensList[T] =
-      new GensList[T]:
-        def gens: List[Gen[T]] = List(a.arbitrary)
-  }
+  def gen: Gen[T] = gens match
+    case List(gen) => gen
+    case gens      => Gen.choose(0, gens.size - 1).flatMap(i => gens(i))
 
-  private object GensList extends LowPriorityGensListGivens:
+private object Gens:
 
-    inline given derived[T](using s: Mirror.SumOf[T]): GensList[T] =
-      val elems = summonAll[Tuple.Map[s.MirroredElemTypes, GensList]]
-      val elemsList = elems.toList.asInstanceOf[List[GensList[T]]]
-      val gs: List[Gen[T]] = elemsList.flatMap(_.gens)
-      new GensList[T]:
-        def gens: List[Gen[T]] = gs
+  def apply[T](gen: Gen[T]): Gens[T] = Gens(List(gen))
 
-  inline private def arbitrarySum[T](s: Mirror.SumOf[T]): Arbitrary[T] =
-    val gens = summonInline[GensList[T]].gens
-    Arbitrary(Gen.choose(0, gens.size - 1).flatMap(i => gens(i)))
-
-  inline private def genTuple[T <: Tuple]: Gen[T] =
+  inline private def tupleInstance[T <: Tuple]: Gens[T] =
     inline erasedValue[T] match
       case _: EmptyTuple =>
-        Gen.const(EmptyTuple.asInstanceOf[T])
+        Gens(Gen.const(EmptyTuple.asInstanceOf[T]))
       case _: (t *: ts) =>
-        for {
-          tVal <- summonInline[Arbitrary[t]].arbitrary
-          tsVal <- genTuple[ts]
-        } yield (tVal *: tsVal).asInstanceOf[T]
+        val gen: Gen[T] =
+          for {
+            tVal <- summonInline[Arbitrary[t]].arbitrary
+            tsVal <- tupleInstance[ts].gen
+          } yield (tVal *: tsVal).asInstanceOf[T]
+        Gens(gen)
+
+  inline def productInstance[T](p: Mirror.ProductOf[T]): Gens[T] =
+    Gens(tupleInstance[p.MirroredElemTypes].gen.map(p.fromProduct(_)))
```

Comment:
Pasting line 40 into the two places where `productInstance` is called could have reduced the total number of inlining steps by one (making this a non-breaking change), but I'd rather keep the code clean.